### DISCUSSION TITLE
Engine Hat - Simulated Fuel Consumption

### DISCUSSION THREAD

**Discussion by sass00** [hatlabs/discussions#66] [Ideas]
Hey guys, 

I've been at this for the day now and i'm slowly giving up :) i've wanted to do a simulated fuel consumption meter based on RPM's i'm getting from my kDigitalInput from my engine hat. In order to do that i decided to edit digital.cpp file, where the RPM based float producer is hiding. Im not good at writing code but im trying to learn.

FloatProducer* ConnectTachoSender(int pin, String name) {
  char config_path[80];
  char sk_path[80];

  snprintf(config_path, sizeof(config_path), "", name.c_str());
  auto tacho_input = new DigitalInputCounter(pin, INPUT, RISING, 500, config_path);

  snprintf(config_path, sizeof(config_path), "/Tacho %s/Revolution Multiplier", name.c_str());
  auto tacho_frequency = new Frequency(kDefaultFrequencyScale, config_path);
  tacho_frequency->connect_to(new Frequency())
    ->connect_to(new MovingAverage(4, 1.0,"/%s Engine Fuel Flow/movingAVG"))
    ->connect_to(new FuelInterpreter("/Engine %s/Fuel/curve"))
    ->connect_to(new SKOutputFloat("propulsion.%s.fuelconsumption", "/%s/ Engine Fuel/skPath"));

  snprintf(config_path, sizeof(config_path), "/Tacho %s/Revolutions SK Path", name.c_str());
  snprintf(sk_path, sizeof(sk_path), "propulsion.%s.revolutions", name.c_str());
  auto tacho_frequency_sk_output = new SKOutputFloat(sk_path, config_path);

  tacho_input
    ->connect_to(tacho_frequency)
    ->connect_to(tacho_frequency_sk_output);

This is what i have made: so the sk path actually shows up in the sk server, but with %S attached (propulsion.%s.fuelconsumption) instead of %s i was hoping for an engine name that is defined in main.cpp (port or stbd). I have done the curveinterpolator at the beggining of the file, but it seems to do the math wrong; 

So when i'm getting 650rpm cca its calculationg about 7lph insted of 2.90-4lph. Can anyone share the idea of what im doing wrong?

**Comment by liuliano** [comment]
1) Why you see propulsion.%s.fuelconsumption literally
SKOutputFloat("propulsion.%s.fuelconsumption", ...) does not do printf-style formatting. It just takes the string as-is, so %s will never be replaced.
You need to snprintf() the final string into a buffer first (same for your MovingAverage config path etc.).
Example pattern:
Copy code
Cpp
char sk_fuel_path[80];
char cfg_fuel_path[80];

snprintf(sk_fuel_path, sizeof(sk_fuel_path),
         "propulsion.%s.fuelconsumption", name.c_str());

snprintf(cfg_fuel_path, sizeof(cfg_fuel_path),
         "/%s Engine Fuel/skPath", name.c_str());

auto fuel_sk = new SKOutputFloat(sk_fuel_path, cfg_fuel_path);
Do the same for:
"/%s Engine Fuel Flow/movingAVG" (you currently pass a %s string but never format it)
any FuelInterpreter("/Engine %s/Fuel/curve") style path (also currently not formatted)
2) Why the fuel math is “way off”
A) Signal K engine revolutions are Hz, not RPM
In Signal K, propulsion.<engine>.revolutions is Hz (revolutions per second), and you multiply by 60 if you want RPM. �
Signal K
So if your curve is defined using RPM values (650, 1000, …) but you feed it Hz (650 RPM ≈ 10.83 Hz), interpolation will be wrong unless you convert.
B) You’re very likely applying Frequency wrong (or twice)
Your chain includes:
Copy code
Cpp
tacho_frequency->connect_to(new Frequency())
But Frequency is meant to convert pulse counts per sample window into Hz (see SensESP rpm example). �
If you feed a Frequency output into another Frequency, the second one is no longer operating on pulse counts — that can easily scale things incorrectly / nonsensically.
Signal K
Typical correct wiring is: DigitalInputCounter (pulses) -> Frequency (Hz) -> (optional smoothing) -> (optional convert to RPM) -> your curve -> SKOutput
SensESP’s example uses one Frequency with a multiplier that accounts for pulses-per-revolution (e.g., 1/97 for a 97-tooth signal). �
Signal K
A corrected structure (conceptually)
If your curve is in RPM, do this:
Copy code
Cpp
auto tacho_frequency = new Frequency(multiplier, "/Tacho .../Revolution Multiplier"); // outputs Hz

tacho_input
  ->connect_to(tacho_frequency)          // Hz
  ->connect_to(new Linear(60.0, 0.0))    // RPM (Hz * 60)
  ->connect_to(new MovingAverage(4, 1.0, moving_avg_cfg))
  ->connect_to(new FuelInterpreter(curve_cfg))
  ->connect_to(new SKOutputFloat(sk_fuel_path, cfg_fuel_path));
Key points:
Only one Frequency.
Convert Hz → RPM (×60) if your curve expects RPM. �
Signal K
Make sure your multiplier matches your signal:
If you get 1 pulse per revolution, multiplier is 1.0.
If you get N pulses per revolution, multiplier should be 1.0 / N. �
Signal K
Quick sanity check for your numbers
650 RPM should be about 650 / 60 = 10.83 Hz.
If you see something like ~21.7 Hz, you might be counting 2 pulses per rev (or you have RISING vs CHANGE issues), so you’d need multiplier = 1.0/2 (or fix the interrupt mode).
If you paste the part of your code where multiplier (or “Revolution Multiplier”) is defined + what your tach signal is (pulses per revolution / teeth count), I can tell you exactly what that multiplier should be and where to place the RPM conversion so the curve hits 2.9–4 lph at ~650 RPM.
